### PR TITLE
feat: rebuild the runtime when a function changes

### PR DESCRIPTION
**Pull request by lance**
Fixes: https://github.com/openshift-cloud-functions/js-function-operator/issues/13

Extract all of the build code into a separate file, and adjust the logic
in the controller so that the ConfigMap is updated and a new TaskRun is
created when the function source changes.

### CODE REVIEW

File: pkg/apis/faas/v1alpha1/jsfunction_types.go [hunk 1]
```diff
@@ -10,9 +10,10 @@ import (
 // JSFunctionSpec defines the desired state of JSFunction
 // +k8s:openapi-gen=true
 type JSFunctionSpec struct {
-	Func    string `json:"func"`
-	Package string `json:"package,omitempty"`
-	Events  bool   `json:"events,omitempty"`
+	Func       string `json:"func"`
+	Deployment int    `json:"deployment"`
```
Comment (slinkydeveloper) [inline pkg/apis/faas/v1alpha1/jsfunction_types.go]: This new parameter sets the incremental index of the "revision" of the src code, right? Why is mandatory? Maybe you don't want to explicit, just go with default 0 and then 1 etc...

Comment (zroubalik) [reply pkg/apis/faas/v1alpha1/jsfunction_types.go]: +1 I have the same question. Could you please explain the idea behind this please?

**Comment by lance** [comment]
@zroubalik @slinkydeveloper I've been thinking about this PR and doing some work on another branch that is a complete reworking of this. I think maybe it would be better to have a `JSFunctionBuild` CRD that is responsible for managing the `PipelineResource` and `PipelineRuns` for a given function. I've got that operator written[1]. Doing this would allow the `JSFunction` operator to simply create the eventing bits, create and update the `ConfigMap`, and create and update the `JSFunctionBuild`  when the function changes. The separation of concerns here is good, and would make everything a little cleaner. Any thoughts on this?

[1] [js-function-build-operator](https://github.com/openshift-cloud-functions/js-function-build-operator/)

**Comment by zroubalik** [comment]
> @zroubalik @slinkydeveloper I've been thinking about this PR and doing some work on another branch that is a complete reworking of this. I think maybe it would be better to have a `JSFunctionBuild` CRD that is responsible for managing the `PipelineResource` and `PipelineRuns` for a given function. I've got that operator written[1]. Doing this would allow the `JSFunction` operator to simply create the eventing bits, create and update the `ConfigMap`, and create and update the `JSFunctionBuild` when the function changes. The separation of concerns here is good, and would make everything a little cleaner. Any thoughts on this?
> 
> [1] [js-function-build-operator](https://github.com/openshift-cloud-functions/js-function-build-operator/)

Sorry, missed this somehow. I like the idea! Will look at the other PR.

**Comment by lance** [comment]
I have removed the `js-function-build-operator` repository and instead added the type and controller to this project. I'm going to close this PR now in favor of https://github.com/openshift-cloud-functions/js-function-operator/pull/20

**Pull request closed without merging**